ostree/ostree_dest.go

```diff
@@ -0,0 +1,246 @@
+
+	size, err := io.Copy(blobFile, tee)
+	if err != nil {
+		return types.BlobInfo{}, err
+	}
+	computedDigest := digester.Digest()
+	if inputInfo.Size != -1 && size != inputInfo.Size {
+		return types.BlobInfo{}, errors.Errorf("Size mismatch when copying %s, expected %d, got %d", computedDigest, inputInfo.Size, size)
+	}
+	if err := blobFile.Sync(); err != nil {
+		return types.BlobInfo{}, err
+	}
+
+	hash := computedDigest.Hex()
+	d.blobs[hash] = &blobToImport{Size : size, Digest : computedDigest, BlobPath : blobPath}
+	return types.BlobInfo{Digest: computedDigest, Size: size}, nil
+}
+
+func (d *ostreeImageDestination) importBlob(blob *blobToImport) (error) {
+	ostreeBranch := fmt.Sprintf("ociimage/%s", blob.Digest.Hex())
+	destinationPath := filepath.Join(encodeOStreeRef(d.ref.image.String()), blob.Digest.Hex(), "root")
+	if err := ensureDirectoryExists(destinationPath); err != nil {
+		return err
+	}
+	defer func() {
+		os.Remove(blob.BlobPath)
+		os.RemoveAll(destinationPath)
+	}()
+
+	// The caller of the function will free this directory
+	err := archive.UntarPath(blob.BlobPath, destinationPath)
+	if err != nil {
+		return err
+	}
+	return exec.Command("ostree", "commit",
+		fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+		fmt.Sprintf("--branch=%s", ostreeBranch),
+		fmt.Sprintf("--tree=dir=%s", destinationPath)).Run()
+}
+
+func (d *ostreeImageDestination) importConfig(blob *blobToImport) (error) {
+	ostreeBranch := fmt.Sprintf("ociimage/%s", blob.Digest.Hex())
+
+	return exec.Command("ostree", "commit",
+		fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+		fmt.Sprintf("--branch=%s", ostreeBranch), filepath.Dir(blob.BlobPath)).Run()
+}
+
+func (d *ostreeImageDestination) HasBlob(info types.BlobInfo) (bool, int64, error) {
+	branch := fmt.Sprintf("ociimage/%s", info.Digest.Hex())
+	output, err := exec.Command("ostree", "show", "--print-metadata-key=docker.size", branch).CombinedOutput()
+	if err != nil {
+		if  bytes.Index(output, []byte("not found")) >= 0 || bytes.Index(output, []byte("No such")) >= 0 {
+			return false, -1, nil
+		} else {
+			return false, -1, err
+		}
+	}
+	size, err := strconv.ParseInt(strings.Trim(string(output), "'\n"), 10, 64)
+	if err != nil {
+		return false, -1, err
+	}
+
+	return true, size, nil
+}
+
+func (d *ostreeImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInfo, error) {
+	return info, nil
+}
+
+func (d *ostreeImageDestination) PutManifest(manifest []byte) error {
+	d.manifest = string(manifest)
+
+	if err := json.Unmarshal(manifest, &d.schema); err != nil {
+		return err
+	}
+
+	manifestPath := d.ref.manifestPath()
+	if err := ensureParentDirectoryExists(manifestPath); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(manifestPath, manifest, 0644)
+}
+
+func (d *ostreeImageDestination) PutSignatures(signatures [][]byte) error {
+	if err := ensureParentDirectoryExists(d.ref.signaturePath(0)); err != nil {
```

Comment:
yes, since the content in the ostree repo cannot be modified once written, we need to store it with the correct SELinux label.  The correct SELinux label is what is assigned once checked-out to `/var/lib/containers/atomic`, we can run skopeo in a temporary directory in `/var/lib/containers/atomic` so that the files are properly labelled before being committed.